### PR TITLE
p2p: refactor and cleanup bootnode relay wiring

### PR DESCRIPTION
**Pull request by corverroos**
- Decouple relays and bootnodes from udpnode.
- Improve gater tests

category: refactor 
ticket: #413

### CODE REVIEW

File: p2p/discovery.go [hunk 1]
```diff
@@ -107,59 +59,10 @@ func NewUDPNode(ctx context.Context, config Config, ln *enode.LocalNode, key *ec
...
-	}, nil
-}
-
-// queryBootnodeENR returns the bootnode ENR via a http GET query to the url.
-//
-// This supports resolving bootnode ENR from known http URLs which is handy
-// when bootnodes are deployed in docker-compose or kubernetes
-//
-// It retries until the context is cancelled.
-func queryBootnodeENR(ctx context.Context, bootnodeURL string, backoff time.Duration) (string, error) {
```
Comment (corverroos) [inline p2p/discovery.go:125]: this is moved to `bootnode.go`

File: p2p/gater_test.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package p2p_test
```
Comment (corverroos) [inline p2p/gater_test.go:16]: made test external and simplified

File: p2p/p2p.go [hunk 3]
```diff
@@ -37,17 +37,9 @@ import (
 	"github.com/obolnetwork/charon/app/z"
 )
 
-// EmptyAdvertisedAddrs defines a p2pcfg.AddrsFactory that does not advertise
-// addresses via libp2p, since we use discv5 for peer discovery.
-var EmptyAdvertisedAddrs = func([]ma.Multiaddr) []ma.Multiaddr { return nil }
```
Comment (corverroos) [inline p2p/p2p.go:42]: this was only used by tests, remove now

File: p2p/peer.go [hunk 4]
```diff
@@ -40,27 +40,6 @@ type Peer struct {
 	Index int
 }
 
-// newRelayPeer returns a new relay peer.
-func newRelayPeer(nodeAddr string) (Peer, error) {
```
Comment (corverroos) [inline p2p/peer.go:44]: not required anymore

File: p2p/discovery.go (hunk 1 above)
Comment (dB2510) [reply p2p/discovery.go:125]: nice

File: app/app.go [hunk 5]
```diff
@@ -169,34 +169,38 @@
...
 
-	connGater, err := p2p.NewConnGater(manifest.PeerIDs(), udpNode.Relays)
+	udpNode, err := p2p.NewUDPNode(conf.P2P, localEnode, p2pKey, bootnodes)
 	if err != nil {
-		return nil, nil, errors.Wrap(err, "connection gater")
+		return nil, nil, err
 	}
 
-	tcpNode, err := p2p.NewTCPNode(conf.P2P, p2pKey, connGater, udpNode, manifest.Peers, p2p.EmptyAdvertisedAddrs)
+	relays, err := p2p.NewRelays(conf.P2P, bootnodes)
```
Comment (dB2510) [inline app/app.go:191]: this can return `nil,nil` why aren't we checking for nil for `relays`?

File: p2p/relay.go [hunk 6]
```diff
@@ -30,6 +31,28 @@
 	"github.com/obolnetwork/charon/app/z"
 )
 
+// NewRelays returns the libp2p circuit relays from bootnodes if enabled.
+func NewRelays(conf Config, bootnodes []*enode.Node) ([]Peer, error) {
+	if !conf.BootnodeRelay {
+		return nil, nil
+	} else if conf.UDPBootManifest {
+		// Relays not supported via manifest bootnodes yet.
+		return nil, nil
```
Comment (dB2510) [inline p2p/relay.go:40]: maybe add a unit test for these scenarios

**Review by dB2510 — APPROVED** [review]
LGTM

File: app/app.go (hunk 5 above)
Comment (corverroos) [reply app/app.go:191]: `nil` is a valid empty slice, so not needed

File: p2p/relay.go (hunk 6 above)
Comment (corverroos) [reply p2p/relay.go:40]: yeah

**Pull request merged**